net/art/table.go

```diff
@@ -156,20 +488,152 @@ func (t *Table[T]) debugSummary() string {
+	// Consider 192.168.1.0/24 and 192.168.0.0/16. commonBits
+	// correctly reports that these prefixes have their first 16 bits
+	// in common. However, in the ART they only share 1 common stride:
+	// they both use the 192.0.0.0/8 strideTable, but 192.168.0.0/16
+	// is stored as 168/8 within that table, and not as 0/0 in the
+	// 192.168.0.0/16 table.
+	//
+	// So, when commonBits matches the length of one of the inputs and
+	// falls on a boundary between strides, the strideTable one
+	// further up from commonBits/8 is the one we need to create,
+	// which means we have to adjust the stride count down by one.
+	if commonBits == minPrefixLen {
+		commonBits--
+	}
+	commonStrides := commonBits / 8
+	lastCommon, err := a.Addr().Prefix(commonStrides * 8)
+	if err != nil {
+		panic(fmt.Sprintf("computePrefixSplit constructing common prefix: %v", err))
+	}
+	if a.Addr().Is4() {
+		aStride = a.Addr().As4()[commonStrides]
+		bStride = b.Addr().As4()[commonStrides]
+	} else {
+		aStride = a.Addr().As16()[commonStrides]
+		bStride = b.Addr().As16()[commonStrides]
+	}
+	return lastCommon, aStride, bStride
+}
+
+// commonBits returns the number of common leading bits of a and b.
+// If the number of common bits exceeds maxBits, it returns maxBits
+// instead.
+func commonBits(a, b netip.Addr, maxBits int) int {
+	if a.Is4() != b.Is4() {
+		panic("commonStrides called with mismatched address families")
+	}
+	var common int
+	// The following implements an old bit-twiddling trick to compute
+	// the number of common leading bits: if you XOR two numbers
+	// together, equal bits become 0 and unequal bits become 1. You
+	// can then count the number of leading zeros (which is a single
+	// instruction on modern CPUs) to get the answer.
+	//
+	// This code is a little more complex than just XOR + count
+	// leading zeros, because IPv4 and IPv6 are different sizes, and
+	// for IPv6 we have to do the math in two 64-bit chunks because Go
+	// lacks a uint128 type.
+	if a.Is4() {
+		aNum, bNum := ipv4AsUint(a), ipv4AsUint(b)
+		common = bits.LeadingZeros32(aNum ^ bNum)
+	} else {
+		aNumHi, aNumLo := ipv6AsUint(a)
+		bNumHi, bNumLo := ipv6AsUint(b)
+		common = bits.LeadingZeros64(aNumHi ^ bNumHi)
+		if common == 64 {
+			common += bits.LeadingZeros64(aNumLo ^ bNumLo)
+		}
+	}
+	if common > maxBits {
+		common = maxBits
+	}
+	return common
+}
+
+// ipv4AsUint returns ip as a uint32.
+func ipv4AsUint(ip netip.Addr) uint32 {
+	bs := ip.As4()
+	return uint32(bs[0])<<24 | uint32(bs[1])<<16 | uint32(bs[2])<<8 | uint32(bs[3])
+}
+
+// ipv6AsUint returns ip as a pair of uint64s.
+func ipv6AsUint(ip netip.Addr) (uint64, uint64) {
+	bs := ip.As16()
+	hi := uint64(bs[0])<<56 | uint64(bs[1])<<48 | uint64(bs[2])<<40 | uint64(bs[3])<<32 | uint64(bs[4])<<24 | uint64(bs[5])<<16 | uint64(bs[6])<<8 | uint64(bs[7])
+	lo := uint64(bs[8])<<56 | uint64(bs[9])<<48 | uint64(bs[10])<<40 | uint64(bs[11])<<32 | uint64(bs[12])<<24 | uint64(bs[13])<<16 | uint64(bs[14])<<8 | uint64(bs[15])
```

Comment:
if you want, encoding/binary.BigEndian.Uint64(bs[:8]) etc should compile down to exactly the same I believe (`binary.BigEndian` is a concrete unexported type specifically so the compiler can inline and not worry about seeing through an interface)

Reply:
I thought from past optimization work that binary.BigEndian specifically did _not_ trigger a bunch of optimizations that were possible (broke inlining or simplification or something?). I remember either Josh or @dsnet filing Go compiler bugs about it?

Also easy to A/B test if I can't find the upstream bug.

Reply:
Couldn't find the bug I was thinking of. Benchmarks inconclusive, some shapes of route table got a bit faster, some got a bit slower, allocs moved around by 0-1% in either direction. I'm not sure how to interpret that, but I'm thinking that means the change is in the noise compared to the variance of my machine while it's running the bench.

Reply:
Having recently investigated `binary.BigEndian` and `binary.LittleEndian`, the code behind `binary.BigEndian` looks basically the same as what you have written there, and then the Go compiler optimizes it into efficient instructions if they exist. It takes a slice instead of an array and then has some magic syntax to ensure that only one length check occurs; here you have an array so can bypass that. I wouldn't be surprised if the resulting assembly is pretty similar with both implementations.

Reply:
Going to switch to encoding/binary. The benchmarks are still kinda inconclusive, with some table shapes getting inexplicably faster/slower under this change, but it's not definitely better or definitely worse, and the readability benefit is probaby worth it.